[inductor] Fix propagating  torch.utils._sympy.functions.Identity in IndexPropagation

yushangdi opened the pull request:
Fixes https://github.com/pytorch/pytorch/issues/160535

Index may contain ` torch.utils._sympy.functions.Identity`. When we call `SymPyOps.index_expr`, if the value is a sympy.Expr with Identity, `TypedExpr(value, dtype)` will fail. So when we unwrap arguments, we expand the sympy expression to unwrap Identity.

Test Plan:
buck run @mode/dev-nosan //caffe2/test/inductor:test_aot_inductor -- -r test_sym_expr_indexing

Rollback Plan:

Differential Revision: D76308640




cc @voznesenskym @penguinwu @EikanWang @jgong5 @Guobing-Chen @XiaobingSuper @zhuhaozhe @blzheng @wenzhe-nrv @jiayisunx @ipiszy @chenyang78 @kadeng @muchulee8 @amjames @chauhang @aakhundov @coconutruben

desertfire reviewed (commented):
Do you plan to land this? If yes, please add a test.

yushangdi commented:
not landing

eellison reviewed (commented):
Can you say more about what the error was previously ? We don't actually want to expand Identity because that might take us into int64 arithmetic in a way that is not handled currently.

eellison reviewed (commented):
Can we expand the Identity only after `if _is_constant(self.expr):` then?

yushangdi committed:
[inductor] Fix propagating  torch.utils._sympy.functions.Identity in IndexPropagation (#155504)

Summary:

Fixes: https://github.com/pytorch/pytorch/issues/160535

Index may contain  torch.utils._sympy.functions.Identity. When we call SymPyOps.index_expr, if the value is a sympy.Expr with Identity, TypedExpr(value, dtype) will fail. So when we unwrap arguments, we expand the sympy expression to unwrap Identity.

Test Plan:
```
buck run mode/dev-nosan //caffe2/test/inductor:test_aot_inductor -- -r test_sym_expr_indexing
```

Rollback Plan:

Differential Revision: D76308640

clee2000 commented:
@pytorchbot merge -f "merged internally"
